### PR TITLE
KFSPTS-12164: Add new-subacct feature to global subacct doc

### PR DESCRIPTION
**Pull request by db483**

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java [hunk 1]
```diff
@@ -351,18 +364,18 @@ protected boolean checkCgCostSharingRules(String subAccountTypeCode) {
...
+        	success &= checkEmptyBOField(KFSPropertyConstants.A21_SUB_ACCOUNT + KFSConstants.DELIMITER + KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT_NUMBER, a21.getCostShareSourceAccountNumber(), getDisplayNameForSubAccountProperty(KFSPropertyConstants.A21_SUB_ACCOUNT + KFSConstants.DELIMITER + KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT_NUMBER));
         }
         
         if (!allFieldsSet && StringUtils.isNotBlank(a21.getCostShareSourceAccountNumber())) {
-        	success &= checkEmptyBOField(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_CHART_OF_ACCOUNTS_CODE, a21.getCostShareChartOfAccountCode(),  getDisplayName(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_CHART_OF_ACCOUNTS_CODE));
+        	success &= checkEmptyBOField(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_CHART_OF_ACCOUNTS_CODE, a21.getCostShareChartOfAccountCode(),  getDisplayNameForSubAccountProperty(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_CHART_OF_ACCOUNTS_CODE));
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java]: another place to use  KFSConstants.DELIMITER in replace of "."

File: src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java [hunk 2]
```diff
@@ -351,18 +364,18 @@ protected boolean checkCgCostSharingRules(String subAccountTypeCode) {
...
         }
 
         a21.refreshReferenceObject(KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT);
         // existence test on Cost Share Account
         if (allFieldsSet) {
             if (ObjectUtils.isNull(a21.getCostShareAccount())) {
-                putFieldError(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT_NUMBER, KFSKeyConstants.ERROR_EXISTENCE, getDisplayName(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT_NUMBER));
+                putFieldError(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT_NUMBER, KFSKeyConstants.ERROR_EXISTENCE, getDisplayNameForSubAccountProperty(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT_NUMBER));
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java]: another place to use  KFSConstants.DELIMITER in replace of "."

File: src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java [hunk 3]
```diff
@@ -371,7 +384,7 @@ protected boolean checkCgCostSharingRules(String subAccountTypeCode) {
         // existence test on Cost Share SubAccount
         if (allFieldsSet && StringUtils.isNotBlank(a21.getCostShareSourceSubAccountNumber())) {
             if (ObjectUtils.isNull(a21.getCostShareSourceSubAccount())) {
-                putFieldError(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_SUB_ACCOUNT_NUMBER, KFSKeyConstants.ERROR_EXISTENCE, getDisplayName(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_SUB_ACCOUNT_NUMBER));
+                putFieldError(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_SUB_ACCOUNT_NUMBER, KFSKeyConstants.ERROR_EXISTENCE, getDisplayNameForSubAccountProperty(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_SUB_ACCOUNT_NUMBER));
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java]: another place to use  KFSConstants.DELIMITER in replace of "."

File: src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java [hunk 4]
```diff
@@ -1,25 +1,30 @@
...
 
 import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Consumer;
 
 import org.apache.commons.lang.StringUtils;
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobal.java]: should this be switched to use .lang3?

File: src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobalNewAccountDetail.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+    public SubAccount getSubAccount() {
+        return subAccount;
+    }
+
+    public void setSubAccount(SubAccount subAccount) {
+        this.subAccount = subAccount;
+    }
+
+    public Integer getContractsAndGrantsAccountResponsibilityIdForRouting() {
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobalNewAccountDetail.java]: This is used for routing, check SubAccountGlobalNewAccountDetail.xml.

File: src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java [hunk 6]
```diff
@@ -2,8 +2,10 @@
 
 import java.util.Collection;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 import org.apache.commons.lang.StringUtils;
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java]: done

**Comment by db483** [comment]
Jay,

Thank you for your review. I have added the changes you have suggested and now this is ready for review again.

File: src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobalNewAccountDetail.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/coa/businessobject/SubAccountGlobalNewAccountDetail.java]: OK, I found it on line 373 of SubAccountGlobalMaintenanceDocument.xml

**Pull request merged**